[Merged by Bors] - feat(geometry/euclidean): angles and some basic lemmas

jsm28 opened the pull request:
Define angles (undirected, between 0 and π, in terms of inner
product), and prove some basic lemmas involving angles, for real inner
product spaces and Euclidean affine spaces.

From the 100-theorems list, this provides versions of
* 04 Pythagorean Theorem,
* 65 Isosceles Triangle Theorem and
* 94 The Law of Cosines, with various existing definitions implicitly providing
* 91 The Triangle Inequality.

jsm28 committed:
feat(geometry/euclidean): angles and some basic lemmas

Define angles (undirected, between 0 and π, in terms of inner
product), and prove some basic lemmas involving angles, for real inner
product spaces and Euclidean affine spaces.

From the 100-theorems list, this provides versions of 4 Pythagorean
Theorem, 65 Isosceles Triangle Theorem and 94 The Law of Cosines (with
various existing definitions implicitly providing 91 The Triangle
Inequality).

It seems reasonable to me for some geometrical results about real
inner product spaces to go in `geometry/euclidean.lean` rather than
`analysis/normed_space/real_inner_product.lean`.  However, some of the
lemmas about real inner product spaces don't involve angles and aren't
particularly geometrical; quite possibly some of the lemmas should
actually go in `analysis/normed_space/real_inner_product.lean`.

Note that there are many other basic geometrical concepts not yet
defined in mathlib, and the results proved are strongly biased to
those that can conveniently be stated and proved with only the
definitions present so far.  For example, all of the following would
be relevant to more complete support for Euclidean geometry, but are
not yet present (and all of them would need many basic results once
added):

* lines and planes (affine subspaces of a given dimension);

* collinearity and coplanarity of points;

* circles and spheres (I expect it will be useful to have a structure
  type that stores the centre, radius and a proof that the radius is
  positive, not just the `sphere` as a set of points defined for
  metric spaces), and tangency;

* n-simplexes (n+1 affine-independent points), and thus triangles as
  2-simplexes;

* area and volume (n-dimensional Lebesgue measure) (and also length of
  a curve / area of a surface);

* betweenness, as in Tarski's axioms (although some lemmas added here
  about angles equal to π are equivalent to corresponding results for
  strict betweenness);

* various other kinds of angles (undirected angles between 0 and 2π,
  using some form of disambiguation for which side of the angle is
  wanted; directed angles modulo 2π, in oriented 2-space, as a
  function of a rotation; directed angles modulo π, in oriented
  2-space, as the rotation angle between two lines, which are often
  the most convenient kind of angle to reduce configuration-dependence
  in proofs; solid angles, in any dimension);

* Cartesian and barycentric frames and coordinates (I suspect it will
  be useful to be able to work with frames for a subspace, not just
  for the whole Euclidean space);

* similarity transformations, and properties of similarities and
  isometries in Euclidean space.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -5,8 +5,10 @@ Author: Joseph Myers.
 -/
 import analysis.normed_space.real_inner_product
 import analysis.normed_space.add_torsor
+import tactic.apply_fun
```
I think we should just add `tactic.apply_fun` to the default tactic imports. @urkud do you agree?

urkud replied:
Why do you ask me? AFAIR I never touched the default tactic imports.

jcommelin replied:
Hmm... sorry. Some connection went wrong in my brain. @digama0 you're the boss. What's the policy with adding stuff to the default tactic imports?

bryangingechen replied:
[`tactic.apply_fun` is already in `src/tactic/default.lean`](https://github.com/leanprover-community/mathlib/blob/master/src/tactic/default.lean#L25). But I don't think we want to `import tactic` in all other files in mathlib, so what exactly do you have in mind?

kim-em replied:
Probably @jcommelin intended to suggest we import `tactic.apply_fun` already in `src/tactic/basic.lean`. That seems very reasonable.

bryangingechen replied:
Ah, that makes sense. But I'm not so sure that's a good idea either. `tactic.apply_fun` imports `tactic.monotonicity`, which in turn imports `algebra.order_functions`, etc.

kim-em replied:
I'll mark this suggestion as resolved (i.e. no change necessary), then.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+
+/-- The undirected angle between two vectors. If either vector is 0,
+this is π/2. -/
+def angle_of_vectors (x y : V) : ℝ := (inner x y / (∥x∥ * ∥y∥)).arccos
```
I think it would be better to put this in a namespace that can be opened, so that most of the time you can simply write `angle`.

jsm28 replied:
Now called `angle` in `namespace inner_product_geometry`.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+end
+
+/-- The angle between the negation of two vectors. -/
+@[simp] lemma angle_of_vectors_of_neg_of_neg (x y : V) :
```
This could become `angle_neg_neg`

jsm28 replied:
Done.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+/-- The angle between two vectors is at most π. -/
+lemma angle_of_vectors_le_pi (x y : V) : angle_of_vectors x y ≤ real.pi :=
+real.arccos_le_pi _
+
```
Somewhere around here I would expect `angle_self` that states `angle x x = 0` (under the assumption that `x ≠ 0`).
Aah, I see that it's somewhere below. :+1:

jsm28 replied:
Renamed to `angle_self`.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+real.arccos_le_pi _
+
+/-- The angle between a vector and the negation of another vector. -/
+lemma angle_of_vectors_of_neg_right (x y : V) :
```
`of_neg_right` would mean that "neg_right" is the description of an assumption. So this could become
`angle_neg_right`
Similar comment applies elsewhere.

jsm28 replied:
Renamed various such lemmas.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+end
+
+/-- The angle between a nonzero vector and its negation. -/
+@[simp] lemma angle_of_vectors_of_neg_self_of_nonzero_right {x : V} (hx : x ≠ 0) :
```
`angle_self_neg_of_nonzero`

jsm28 replied:
Renamed (and `angle_neg_self_of_nonzero` likewise).

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+by rw [angle_of_vectors_comm, angle_of_vectors_of_neg_self_of_nonzero_right hx]
+
+/-- The angle between a vector and a positive multiple of a vector. -/
+@[simp] lemma angle_of_vectors_of_pos_smul_right (x y : V) {r : ℝ} (hr : 0 < r) :
```
`angle_smul_right_of_pos`

jsm28 replied:
Renamed various such lemmas.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+by rw [angle_of_vectors_comm, angle_of_vectors_of_neg_smul_right y x hr, angle_of_vectors_comm]
+
+/-- The cosine of the angle between two vectors. -/
+lemma cos_angle_of_vectors (x y : V) :
```
I think it would make sense to prove this lemma directly after the definition of `angle`

jsm28 replied:
Moved up.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+vectors and a third vector add to π. -/
+lemma angle_of_vectors_add_angle_of_vectors_eq_pi_of_angle_of_vectors_eq_pi {x y : V} (z : V)
+    (h : angle_of_vectors x y = real.pi) :
+  angle_of_vectors x z + angle_of_vectors y z = real.pi :=
```
I know it doesn't matter for the maths, but would `angle z y` for the second summand be slightly easier in applications? You probably know this best...

jsm28 replied:
I gave this one in symmetric form, but I suspect what's easier for applications is usually directed angles not these undirected angles at all (directed angles are more complicated to define, first you need to define orientations of finite-dimensional vector spaces / subspaces over ordered fields = equivalence classes of bases where two bases are equivalent if the map between them has positive determinant).

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -5,8 +5,10 @@ Author: Joseph Myers.
+import tactic.apply_fun
 
 noncomputable theory
+open_locale classical
```
There should be notation for `pi` available. I don't know how it's set up at the moment. Maybe `open_locale real` will do the job. If it doesn't we should fix that.

urkud replied:
``localized "notation `π` := real.pi" in real`` goes right after definition of `real.pi`.

jsm28 replied:
Now using π everywhere.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+end
+
+/-- Pythagorean theorem, subtracting vectors, if-and-only-if vector angle form. -/
+lemma norm_sub_square_eq_norm_square_add_norm_square_iff_angle_eq_pi_div_two (x y : V) :
```
These statements are useful! But if I want to rewrite the equality, I need to write
```
(norm_sub_square_eq_norm_square_add_norm_square_iff_angle_eq_pi_div_two x y).mpr h
```
which is quite a mouthful.
Could you please add
`norm_sub_square_eq_norm_square_add_norm_square` (and a version with a `'`) that take the assumption on the angle/inner as hypothesis?

jsm28 replied:
Added four such variants of Pythagoras (for vectors, so there are now eight variants for vectors and one for points; once triangles are defined there will need to be at least one for triangles as well).

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+    (h : angle_of_vectors x (x - y) = angle_of_vectors y (y - x))
+    (hpi : angle_of_vectors x y ≠ real.pi) : ∥x∥ = ∥y∥ :=
+begin
+  replace h := real.arccos_inj (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x (x - y))).1
```
I would put the hypothesis on a newline, indented by two spaces, and all the others below it, like you did now. This keeps all the line lengths a bit shorter.

jsm28 replied:
Adjusted formatting.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+`p3`. If either of those points equals `p2`, this is π/2. -/
+def angle_of_points (p1 p2 p3 : P) : ℝ := angle_of_vectors (p1 -ᵥ p2 : V) (p3 -ᵥ p2)
+
+notation `∠` := angle_of_points
```
I don't think this should be global notation.
You could add `local ` in front of this line, to restrict it to this file.
Alternatively, you can make it `localized`, so that writing `open_locale euclidean_geometry` (or something) will enable the notation.

jsm28 replied:
Now defined as `localized`.

jcommelin commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+
+/-- The undirected angle at `p2` between the line segments to `p1` and
+`p3`. If either of those points equals `p2`, this is π/2. -/
+def angle_of_points (p1 p2 p3 : P) : ℝ := angle_of_vectors (p1 -ᵥ p2 : V) (p3 -ᵥ p2)
```
If you put this in a different namespace, you could call this `angle` again.

jsm28 replied:
Now also called `angle`. (For directed angles modulo 2π the `angle` type in `analysis/special_functions/trigonometric.lean` will come into play as well.)

jcommelin reviewed (commented):
A lot of interesting stuff! Thanks a lot!
I left a bunch of comments. (sometimes a bit brief, but they are certainly not meant harsh)
Mostly about naming things.

jsm28 committed:
Renamings and other changes from review.

jsm28 committed:
Move cos_angle up further.

jsm28 committed:
Add more variants of Pythagoras as per review.

urkud commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +36,642 @@ normed_add_torsor V P
+
+/-- The undirected angle between two vectors. If either vector is 0,
+this is π/2. -/
+def angle (x y : V) : ℝ := (inner x y / (∥x∥ * ∥y∥)).arccos
```
I don't think that using dot notation for `arccos`, `cos` etc is a good idea.

kim-em replied:
For now the pretty printer outputs this, but we're working on preventing that. :-) https://github.com/leanprover-community/lean/issues/289

jsm28 replied:
Changed to avoid it for `sin`, `cos` and `arccos`. I think I was assuming the pretty printer's output represented the preferred notation.

urkud commented on src/geometry/euclidean.lean:
```diff
@@ -33,3 +36,642 @@ normed_add_torsor V P
+norms, has absolute value 1 if and only if they are nonzero and one is
+a multiple of the other. One form of equality case for Cauchy-Schwarz. -/
+lemma abs_inner_div_norm_mul_norm_eq_one_iff (x y : V) :
+  abs (inner x y / (∥x∥ * ∥y∥)) = 1 ↔ (x ≠ 0 ∧ ∃ (r : ℝ), r ≠ 0 ∧ y = r • x) :=
```
* Lemmas before the definition of `angle` can go to `analysis.normed_space.real_inner_product`;
* The length of this proof makes me sad. It would be better to merge the proof of this with the proof of Cauchy-Schwarz. I can look into this tomorrow.

jsm28 replied:
Moved those lemmas. There are also four lemmas with versions of Pythagoras that appear after the definition of `angle` but don't depend on `angle` in any way, should those also move to `analysis.normed_space.real_inner_product`? Feel free to improve proofs / construct a version of Cauchy-Schwarz that proves the required equality case at the same time as the main result.

jsm28 replied:
Checking again on the previous question: should the versions of Pythagoras that don't depend on `angle` also move to `analysis.normed_space.real_inner_product`?

jcommelin replied:
@jsm28 Yes, please move those. After that, we can kick this on the queue. The refactor can also be done in a follow up PR.

jsm28 replied:
Moved those lemmas.

jsm28 committed:
Avoid dot notation for sin, cos, arccos.

jsm28 committed:
Move some lemmas to real_inner_product.

jsm28 committed:
Add back a blank line.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into euclidean-geometry-1

jsm28 committed:
Add lemmas expressing inner product in terms of norm.

Sometimes useful in geometry proofs, although not used in the
particular proofs in this PR.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into euclidean-geometry-1

jsm28 committed:
Update for renamings.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into euclidean-geometry-1

jsm28 committed:
Move versions of Pythagoras for inner product to real_inner_product.lean.

jsm28 committed:
Versions of Pythagoras taking a hypothesis can have implicit arguments.

jcommelin reviewed (approved):
Thanks :tada:

bors merge
